File: add.go
package main

import (
	"fmt"
	"github.com/satori/go.uuid"
	"github.com/urfave/cli"
	"gopkg.in/ldap.v3"
	"log"
	"reflect"
)

type AddJob struct {
	BaseJob
	uuid bool
}

var addFlags = []cli.Flag{
	cli.BoolFlag{
		Name:  "uuid",
		Usage: "use UUID",
	},
}

func Add(c *cli.Context) error {
	runBenchmark(c, reflect.TypeOf(AddJob{}))
	return nil
}

func (job *AddJob) Prep(c *cli.Context) bool {
	if job.GetVerbose() >= 1 {
		log.Printf("worker[%d]: prepare\n", job.wid)
	}
	job.uuid = c.Bool("uuid")
	err := job.conn.Bind(c.String("D"), c.String("w"))
	if err != nil {
		log.Fatal("bind error: ", err)
		return false
	}
	return true
}

func (job *AddJob) Request() bool {
	var cn string
	id, err := uuid.NewV1()
	if job.uuid {
		cn = id.String()
	} else {
		cn = fmt.Sprintf("%d-%d", job.wid, job.count)
	}
	dn := fmt.Sprintf("cn=%s,%s", cn, job.baseDN)
	attrs := []ldap.Attribute{
		ldap.Attribute{"objectClass", []string{"person"}},
		ldap.Attribute{"cn", []string{cn}},
		ldap.Attribute{"sn", []string{"sn"}},
		ldap.Attribute{"userPassword", []string{"secret"}},
	}
	req := ldap.AddRequest{
		DN:         dn,
		Attributes: attrs,
	}
	err = job.conn.Add(&req)
	if err != nil {
		log.Printf("add error: %s", err)
		return false
	}
	return true
}

File: src/add.rs
use async_trait::async_trait;
use clap::Args;
use std::time::Instant;
use uuid::Uuid;

use crate::lb::{BaseJob, CommonArgs, HasCommonArgs, Job};

#[derive(Debug, Args, Clone)]
pub struct AddArgs {
    #[command(flatten)]
    pub common: CommonArgs,

    /// Use UUID for cn attribute
    #[arg(long)]
    pub uuid: bool,

    /// Base DN for entries
    #[arg(short = 'b', long, default_value = "dc=example,dc=com")]
    pub base_dn: String,
}

impl HasCommonArgs for AddArgs {
    fn common(&self) -> &CommonArgs {
        &self.common
    }
}

pub struct AddJob {
    base: BaseJob,
    args: AddArgs,
}

#[async_trait]
impl Job for AddJob {
    type Args = AddArgs;

    fn new(tid: usize, args: &Self::Args) -> Self {
        AddJob {
            base: BaseJob::new(tid),
            args: args.clone(),
        }
    }

    fn args(&self) -> &Self::Args {
        &self.args
    }

    fn base(&mut self) -> &mut BaseJob {
        &mut self.base
    }

    async fn prepare(&mut self) -> bool {
        let common = self.args.common();
        self.base.bind(&common.bind_dn, &common.bind_pw).await
    }

    async fn request(&mut self) -> bool {
        let cn = if self.args.uuid {
            Uuid::new_v4().to_string()
        } else {
            format!("{}", self.base.start_index + self.base.count)
        };

        let dn = format!("cn={},{}", cn, self.args.base_dn);
        let tid_str = self.base.tid.to_string();

        let attrs = vec![
            ("objectClass", vec!["person"].into_iter().collect()),
            ("cn", vec![cn.as_str()].into_iter().collect()),
            ("sn", vec![tid_str.as_str()].into_iter().collect()),
            ("userPassword", vec!["secret"].into_iter().collect()),
        ];

        if let Some(ref mut ldap) = self.base.ldap {
            let start_time = Instant::now();
            let result = ldap.add(&dn, attrs).await;
            let duration = start_time.elapsed();

            self.base.record_latency(duration.as_micros() as u64);

            match result {
                Ok(res) => res.rc == 0,
                Err(e) => {
                    eprintln!("add error: {}", e);
                    false
                }
            }
        } else {
            false
        }
    }
}
